### PR TITLE
feat: gas price oracle

### PR DESCRIPTION
**Pull request by Rjected**
This is a port of the [go-ethereum `Oracle`](https://github.com/ethereum/go-ethereum/blob/master/eth/gasprice/gasprice.go). Currently it is only integrated into `eth_gasPrice` (not `eth_feeHistory`), and tries to be as compatible as possible with the go implementation.

This is also modelled as a service, similar to the `EthStateCache`, which might be a bit overkill given its current capabilities, but it might be fine for now.

TODO:
 - [x] Remove `expect`s - still need to check what geth does in some edge cases and create error variants
 - [ ] Testing - It would be nice to have good rust unit tests, as I'm not sure to what extent hive covers this.

### CODE REVIEW

**Comment by Rjected** [comment]
need to fix tests

File: crates/rpc/rpc/src/eth/gas_oracle.rs [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+        // of prices
+        //
+        // we only return more than check_block blocks' worth of prices if one or more return empty
+        // transactions
+        let mut current_block = header.number;
+        let mut results = Vec::new();
+        let mut populated_blocks = 0;
+
+        // we only check a maximum of 2 * max_block_history
+        for _ in 0..self.oracle_config.max_block_history * 2 {
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/gas_oracle.rs:201]: this looks wasteful to me to fetch the entire history range again.

**Review by mattsse — CHANGES_REQUESTED** [review]
I have some concerns about this.

this currently behaves exactly like a regular `Mutex<U256>` where the value is recomputed from scratch for every new block.

But, I guess having this an extra service is fine, this way we can batch process multiple requests at once, but this entire thing can likely also be a Mutex wrapper 
 
Reading the entire range on every new block seems too expensive, needs caching

File: crates/rpc/rpc/src/eth/gas_oracle.rs (hunk 1 above)
Comment (Rjected) [reply crates/rpc/rpc/src/eth/gas_oracle.rs:113]: `get_block_values` should use the `EthStateCache` now

**Comment by Rjected** [comment]
There is a bug in the math somewhere, and it is still very slow on the first run - looking into it. Otherwise this is good for review

**Comment by Rjected** [comment]
took too long to find this bug 🤦‍♂️
```diff
        if let Some(max_price) = self.oracle_config.max_price {
-            price = max_price;
+            if price > max_price {
+                price = max_price;
+            }
        }
```
The `cast gas-price` responses and values produced by `get_block_values` are now consistent between reth and geth for the tip that I tested

**Comment by Rjected** [comment]
I have one main thing outstanding that I'd like feedback on - we don't expect any sender recovery to fail so what should we do about this case in the filter?
https://github.com/paradigmxyz/reth/blob/02f2415101be03e6c46908c89d3240f312b97046/crates/rpc/rpc/src/eth/gas_oracle.rs#L208-L209

File: crates/rpc/rpc/src/eth/api/mod.rs [hunk 2]
```diff
@@ -238,4 +257,6 @@ struct EthApiInner<Client, Pool, Network> {
     signers: Vec<Box<dyn EthSigner>>,
     /// The async cache frontend for eth related data
     eth_cache: EthStateCache,
+    /// The async gas oracle frontend for gas price suggestions
+    gas_oracle: GasPriceOracle<Client>,
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/mod.rs:261]: this is fine, 
I don't think we need more flexibility here since gas sampling will be similar for all networks?
if not we can easily replace this with a trait and box this.

File: crates/rpc/rpc/src/eth/gas_oracle.rs [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+    /// Suggests a gas price estimate based on recent blocks, using the configured percentile.
+    pub async fn suggest_tip_cap(&self) -> EthResult<U256> {
+        let header_hash = self
+            .client
+            .block_hash_for_id(BlockId::Number(BlockNumberOrTag::Latest))?
+            .expect("a latest header should always exist");
+        let header = self
+            .client
+            .header_by_hash_or_number(BlockHashOrNumber::Hash(header_hash))?
+            .ok_or(GeneralError::Provider(ProviderError::BlockHash { block_hash: header_hash }))?;
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/gas_oracle.rs:104]: we have UnknownBlockNumber for these cases. 
let's replace the except and the `BlockHash` error with that.

https://github.com/paradigmxyz/reth/blob/1776a0c8e0da24532a7d3ccca0ac1e189da344ee/crates/rpc/rpc/src/eth/error.rs#L26-L27

File: crates/rpc/rpc/src/eth/gas_oracle.rs [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+            return Ok(last_price.price)
+        }
+
+        // if all responses are empty, then we can return a maximum of 2*check_block blocks' worth
+        // of prices
+        //
+        // we only return more than check_block blocks' worth of prices if one or more return empty
+        // transactions
+        let mut current_block = header.number;
+        let mut results = Vec::new();
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/gas_oracle.rs:119]: can we pre allocate here, or doesn't really matter?

File: crates/rpc/rpc/src/eth/gas_oracle.rs [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+            None => return Ok(None),
+        };
+
+        // check the cache
+        let block = match self.cache.get_block(block_hash).await? {
+            Some(block) => block,
+            None => match self.client.block_by_number(block_num)? {
+                Some(block) => block,
+                None => return Ok(None),
+            },
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/gas_oracle.rs:186]: this is redundant, because the cache service would have found the block if it exists.

File: crates/rpc/rpc/src/eth/gas_oracle.rs [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+        // of prices
+        //
+        // we only return more than check_block blocks' worth of prices if one or more return empty
+        // transactions
+        let mut current_block = header.number;
+        let mut results = Vec::new();
+        let mut populated_blocks = 0;
+
+        // we only check a maximum of 2 * max_block_history
+        for _ in 0..self.oracle_config.max_block_history * 2 {
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/gas_oracle.rs:123]: we have the latest header, right?
and the range (history * 2),
we could convert this loop so that it tries to traverse the chain by parent hash, but falls back to number if we were unable to find a block with the given hash.

this way we don't need to convert number to hash again

**Review by mattsse — CHANGES_REQUESTED** [review]
nice work!

smol nit and suggestion re block loop

while this entire operation is quite expensive, it is really just once per block.

we could improve performance a bit if the oracle would cache the block range itself and doesn't need to send a lot of messages to the cache, which would cut down on number to hash

File: crates/rpc/rpc/src/eth/gas_oracle.rs (hunk 4 above)
Comment (Rjected) [reply crates/rpc/rpc/src/eth/gas_oracle.rs:114]: Not sure how much we should pre allocate, because we don't know how many values `get_block_values` will return, or exactly how many times we will call it

File: crates/rpc/rpc/src/eth/gas_oracle.rs (hunk 6 above)
Comment (Rjected) [reply crates/rpc/rpc/src/eth/gas_oracle.rs:118]: just converted to traversing entirely by hash

File: crates/rpc/rpc/src/eth/gas_oracle.rs (hunk 5 above)
Comment (Rjected) [reply crates/rpc/rpc/src/eth/gas_oracle.rs]: no longer necessary since the move to traversing by parent hash

File: crates/rpc/rpc/src/eth/gas_oracle.rs [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+        block_hash: H256,
+        limit: usize,
+    ) -> EthResult<Option<(H256, Vec<U256>)>> {
+        // check the cache
+        let block = match self.cache.get_block(block_hash).await? {
+            Some(block) => block,
+            None => match self.client.block_by_hash(block_hash)? {
+                Some(block) => block,
+                None => return Ok(None),
+            },
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/gas_oracle.rs:178]: this branch is not needed, because if the cache is not able to find it (uses the same `block_by_hash` internally) then this won't either

File: crates/rpc/rpc/src/eth/gas_oracle.rs [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+        let mut current_hash = header_hash;
+        let mut results = Vec::new();
+        let mut populated_blocks = 0;
+
+        // we only check a maximum of 2 * max_block_history
+        for _ in 0..self.oracle_config.max_block_history * 2 {
+            let (parent_hash, block_values) = self
+                .get_block_values(current_hash, SAMPLE_NUMBER as usize)
+                .await?
+                .ok_or(EthApiError::UnknownBlockNumber)?;
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/gas_oracle.rs:122]: this can happen if the chain is not long enough yet, so `max_block_history` longer than chain length.

so we should probably cap this with header number

`self.oracle_config.max_block_history * 2`

however, this isn't really an issue because this edge case naturally goes away

**Review by mattsse — APPROVED** [review]
lgtm, pending suggestion re edge case with short chain

File: crates/rpc/rpc/src/eth/gas_oracle.rs (hunk 8 above)
Comment (Rjected) [reply crates/rpc/rpc/src/eth/gas_oracle.rs:131]: makes sense, bounded by block num

File: crates/rpc/rpc/src/eth/gas_oracle.rs (hunk 7 above)
Comment (Rjected) [reply crates/rpc/rpc/src/eth/gas_oracle.rs]: ah right, removed the branch

File: crates/rpc/rpc-builder/tests/it/http.rs [hunk 9]
```diff
@@ -97,8 +97,8 @@ where
     EthApiClient::send_transaction(client, transaction_request).await.unwrap_err();
     EthApiClient::hashrate(client).await.unwrap();
     EthApiClient::submit_hashrate(client, U256::default(), H256::default()).await.unwrap();
-    EthApiClient::gas_price(client).await.unwrap();
-    EthApiClient::max_priority_fee_per_gas(client).await.unwrap();
+    EthApiClient::gas_price(client).await.unwrap_err();
+    EthApiClient::max_priority_fee_per_gas(client).await.unwrap_err();
```
Comment (gakonst) [inline crates/rpc/rpc-builder/tests/it/http.rs:101]: what happened here?

**Review by gakonst — COMMENTED** [review]
Looks fine - IMO this is lacking a ton of tests. I am OK with the fact that we merged it, but @Rjected will need to come back to add more tests.

File: crates/rpc/rpc-builder/tests/it/http.rs (hunk 9 above)
Comment (Rjected) [reply crates/rpc/rpc-builder/tests/it/http.rs:101]: Since we wanted to bound the number of requests to `get_block_values` by `min(blocknum, configured_max)`, this commit https://github.com/paradigmxyz/reth/pull/2600/commits/bcdc4048131b2043a461176803e637a2db57177e changed the call for the latest header from this:
```rust
        let header_hash = self
            .client
            .block_hash_for_id(BlockId::Number(BlockNumberOrTag::Latest))?
```
to this:
```rust
        let block = self
            .client
            .block_by_id(BlockId::Number(BlockNumberOrTag::Latest))?
```
When `Latest` is passed to `block_hash_for_id`, it uses the `BlockNumProvider::chain_info()` method:
https://github.com/paradigmxyz/reth/blob/6e15d72e7db698ed8ac8eb6182bd013219012715/crates/storage/provider/src/traits/block_id.rs#L70-L77

The `NoopProvider` implementation uses a default value for this, which always exists, so the call never fails:
https://github.com/paradigmxyz/reth/blob/6e15d72e7db698ed8ac8eb6182bd013219012715/crates/storage/provider/src/test_utils/noop.rs#L31-L36

The default value for the `ChainInfo` struct is the same as the default hash for `GasPriceOracleResult`, so we used to return early and did not error:
https://github.com/paradigmxyz/reth/blob/cc601dbe9f1451b39caa96f6cd35ffe0699dae35/crates/rpc/rpc/src/eth/gas_oracle.rs#L106-L109

The new code uses `block_by_id`, which returns `Ok(None)`, and ends up in an error `UnknownBlockNumber`:
https://github.com/paradigmxyz/reth/blob/6e15d72e7db698ed8ac8eb6182bd013219012715/crates/storage/provider/src/test_utils/noop.rs#L65-L67
